From 18be004ad18b8a33e5854c08ab2ea023f34890b6 Mon Sep 17 00:00:00 2001
From: Vasilis Liaskovitis <vliaskovitis@suse.com>
Date: Sun, 18 Oct 2020 18:32:17 +0200
Subject: [PATCH 10/10] [PATCH 10/11] modesetting: add TearFree support

All this work to get there, now enjoy your tear-free desktop!

Buglink: https://gitlab.freedesktop.org/xorg/xserver/-/issues/244
Original MR: https://gitlab.freedesktop.org/xorg/xserver/-/merge_requests/24
Signed-off-by: Martin Peres <martin.peres@linux.intel.com>
---
 hw/xfree86/drivers/modesetting/driver.c       | 31 ++++++++--
 hw/xfree86/drivers/modesetting/driver.h       |  2 +
 .../drivers/modesetting/drmmode_display.c     | 26 ++++++--
 .../drivers/modesetting/drmmode_display.h     |  4 ++
 .../drivers/modesetting/modesetting.man       |  8 +++
 hw/xfree86/drivers/modesetting/pageflip.c     | 61 +++++++++++++++++++
 6 files changed, 123 insertions(+), 9 deletions(-)

diff --git a/hw/xfree86/drivers/modesetting/driver.c b/hw/xfree86/drivers/modesetting/driver.c
index b837ac012..ce4cbf791 100644
--- a/hw/xfree86/drivers/modesetting/driver.c
+++ b/hw/xfree86/drivers/modesetting/driver.c
@@ -134,6 +134,7 @@ static const OptionInfoRec Options[] = {
     {OPTION_DOUBLE_SHADOW, "DoubleShadow", OPTV_BOOLEAN, {0}, FALSE},
     {OPTION_ATOMIC, "Atomic", OPTV_BOOLEAN, {0}, FALSE},
     {OPTION_SHADOW_PRIMARY, "ShadowPrimary", OPTV_BOOLEAN, {0}, FALSE},
+    {OPTION_TEARFREE, "TearFree", OPTV_BOOLEAN, {0}, FALSE},
     {-1, NULL, OPTV_NONE, {0}, FALSE}
 };
 
@@ -529,6 +530,13 @@ ms_update_scanout_damages(ScrnInfoPtr scrn, DamagePtr damage)
         RegionUnion(&drmmode_crtc->shadow_nonrotated->screen_damage,
                     &drmmode_crtc->shadow_nonrotated->screen_damage,
                     &pixregion);
+
+        /* also update the shadow backbuffer's damages if available */
+        if (drmmode_crtc->shadow_nonrotated_back) {
+            RegionUnion(&drmmode_crtc->shadow_nonrotated_back->screen_damage,
+                        &drmmode_crtc->shadow_nonrotated_back->screen_damage,
+                        &pixregion);
+        }
     }
 }
 
@@ -722,16 +730,22 @@ ms_schedule_all_scanout_buffers_updates(ScreenPtr pScreen)
     for (c = 0; c < xf86_config->num_crtc; c++) {
         xf86CrtcPtr crtc = xf86_config->crtc[c];
         drmmode_crtc_private_ptr drmmode_crtc = crtc->driver_private;
+        drmmode_ptr drmmode = drmmode_crtc->drmmode;
 
         /* do not do any copies for inactive or rotated displays */
         if (!crtc->active || !drmmode_crtc->shadow_nonrotated)
             continue;
 
-        /* schedule an async update, or revert to a synchronous copy on failure */
-        if (!drmmode_scanout_buffer_update_schedule(crtc,
-                                                    drmmode_crtc->shadow_nonrotated)) {
-            drmmode_update_scanout_buffer(crtc,
-                                          drmmode_crtc->shadow_nonrotated);
+        /* do an async update, unless we are using tearfree or if it failed */
+        if (drmmode->tear_free) {
+            ms_do_tearfree_flip(crtc);
+        } else if (!drmmode_scanout_buffer_update_schedule(crtc,
+                                                           drmmode_crtc->shadow_nonrotated)) {
+                /* since we failed the asynchronous update, fallback to the
+                 * synchronous one
+                 */
+                drmmode_update_scanout_buffer(crtc,
+                                              drmmode_crtc->shadow_nonrotated);
         }
     }
 }
@@ -1049,6 +1063,13 @@ PreInit(ScrnInfoPtr pScrn, int flags)
         xf86DrvMsg(pScrn->scrnIndex, X_INFO, "ShadowPrimary enabled\n");
     }
 
+    ms->drmmode.tear_free =
+            xf86ReturnOptValBool(ms->drmmode.Options, OPTION_TEARFREE,
+                                 FALSE);
+    if (ms->drmmode.tear_free) {
+        xf86DrvMsg(pScrn->scrnIndex, X_INFO, "TearFree enabled\n");
+    }
+
     ms->cursor_width = 64;
     ms->cursor_height = 64;
     ret = drmGetCap(ms->fd, DRM_CAP_CURSOR_WIDTH, &value);
diff --git a/hw/xfree86/drivers/modesetting/driver.h b/hw/xfree86/drivers/modesetting/driver.h
index 1412fe7e7..2ea72e385 100644
--- a/hw/xfree86/drivers/modesetting/driver.h
+++ b/hw/xfree86/drivers/modesetting/driver.h
@@ -53,6 +53,7 @@ typedef enum {
     OPTION_DOUBLE_SHADOW,
     OPTION_ATOMIC,
     OPTION_SHADOW_PRIMARY,
+    OPTION_TEARFREE,
 } modesettingOpts;
 
 typedef struct
@@ -184,4 +185,5 @@ Bool ms_do_pageflip(ScreenPtr screen,
 
 #endif
 
+void ms_do_tearfree_flip(xf86CrtcPtr crtc);
 int ms_flush_drm_events(ScreenPtr screen);
diff --git a/hw/xfree86/drivers/modesetting/drmmode_display.c b/hw/xfree86/drivers/modesetting/drmmode_display.c
index 962994f73..aebab3ed7 100644
--- a/hw/xfree86/drivers/modesetting/drmmode_display.c
+++ b/hw/xfree86/drivers/modesetting/drmmode_display.c
@@ -1642,7 +1642,7 @@ drmmode_need_shadow_scanout(xf86CrtcPtr crtc) {
     if (crtc->rotation != RR_Rotate_0)
         return FALSE;
 
-    if (drmmode->shadow_primary)
+    if (drmmode->shadow_primary || drmmode->tear_free)
         return TRUE;
 
     /* check if the current screen size exceeds the driver's limits */
@@ -1683,6 +1683,7 @@ drmmode_set_mode_major(xf86CrtcPtr crtc, DisplayModePtr mode,
     drmmode_crtc_private_ptr drmmode_crtc = crtc->driver_private;
     drmmode_ptr drmmode = drmmode_crtc->drmmode;
     drmmode_shadow_scanout_ptr saved_shadow_nonrotated;
+    drmmode_shadow_scanout_ptr saved_shadow_nonrotated_back;
     int saved_x, saved_y;
     Rotation saved_rotation;
     DisplayModeRec saved_mode;
@@ -1695,6 +1696,7 @@ drmmode_set_mode_major(xf86CrtcPtr crtc, DisplayModePtr mode,
     saved_y = crtc->y;
     saved_rotation = crtc->rotation;
     saved_shadow_nonrotated = drmmode_crtc->shadow_nonrotated;
+    saved_shadow_nonrotated_back = drmmode_crtc->shadow_nonrotated_back;
 
     if (mode) {
         crtc->mode = *mode;
@@ -1709,16 +1711,27 @@ drmmode_set_mode_major(xf86CrtcPtr crtc, DisplayModePtr mode,
         crtc->funcs->gamma_set(crtc, crtc->gamma_red, crtc->gamma_green,
                                crtc->gamma_blue, crtc->gamma_size);
 
+        drmmode_crtc->shadow_nonrotated = NULL;
+        drmmode_crtc->shadow_nonrotated_back = NULL;
+
         if (drmmode_need_shadow_scanout(crtc)) {
             drmmode_crtc->shadow_nonrotated = drmmode_shadow_scanout_create(crtc);
             if (!drmmode_crtc->shadow_nonrotated) {
                 xf86DrvMsg(crtc->scrn->scrnIndex, X_ERROR,
-                           "drmmode_scanout_shadow_init failed!\n");
+                           "drmmode_scanout_shadow_init failed for the frontbuffer!\n");
+                ret = FALSE;
+                goto done;
+            }
+        }
+
+        if (drmmode->tear_free) {
+            drmmode_crtc->shadow_nonrotated_back = drmmode_shadow_scanout_create(crtc);
+            if (!drmmode_crtc->shadow_nonrotated_back) {
+                xf86DrvMsg(crtc->scrn->scrnIndex, X_ERROR,
+                           "drmmode_scanout_shadow_init failed for the backbuffer!\n");
                 ret = FALSE;
                 goto done;
             }
-        } else {
-            drmmode_crtc->shadow_nonrotated = NULL;
         }
 
         can_test = drmmode_crtc_can_test_mode(crtc);
@@ -1732,6 +1745,7 @@ drmmode_set_mode_major(xf86CrtcPtr crtc, DisplayModePtr mode,
 
         /* We cannot fail anymore, free the previous scanout buffer */
         drmmode_scanout_destroy(crtc, saved_shadow_nonrotated);
+        drmmode_scanout_destroy(crtc, saved_shadow_nonrotated_back);
 
         if (crtc->scrn->pScreen)
             xf86CrtcSetScreenSubpixelOrder(crtc->scrn->pScreen);
@@ -1774,7 +1788,9 @@ drmmode_set_mode_major(xf86CrtcPtr crtc, DisplayModePtr mode,
          * restoring the previous values
          */
         drmmode_scanout_destroy(crtc, drmmode_crtc->shadow_nonrotated);
+        drmmode_scanout_destroy(crtc, drmmode_crtc->shadow_nonrotated_back);
         drmmode_crtc->shadow_nonrotated = saved_shadow_nonrotated;
+        drmmode_crtc->shadow_nonrotated_back = saved_shadow_nonrotated_back;
     } else
         crtc->active = TRUE;
 
@@ -4180,12 +4196,14 @@ drmmode_free_bos(ScrnInfoPtr pScrn, drmmode_ptr drmmode)
         dumb_bo_destroy(drmmode->fd, drmmode_crtc->cursor_bo);
 
         drmmode_scanout_destroy(crtc, drmmode_crtc->shadow_nonrotated);
+        drmmode_scanout_destroy(crtc, drmmode_crtc->shadow_nonrotated_back);
 
         /* HACK: make sure the shadow buffers are NULL because, when reseting, X
          * does not re-allocate the crtc structures, which keeps stale pointers
          * and leads to some use-after-free.
          */
         drmmode_crtc->shadow_nonrotated = NULL;
+        drmmode_crtc->shadow_nonrotated_back = NULL;
     }
 }
 
diff --git a/hw/xfree86/drivers/modesetting/drmmode_display.h b/hw/xfree86/drivers/modesetting/drmmode_display.h
index f47b05208..3b3c2a528 100644
--- a/hw/xfree86/drivers/modesetting/drmmode_display.h
+++ b/hw/xfree86/drivers/modesetting/drmmode_display.h
@@ -107,6 +107,9 @@ typedef struct {
     /** force creating a per-crtc framebuffer **/
     Bool shadow_primary;
 
+    /** Is Option "TearFree" enabled? */
+    Bool tear_free;
+
     /** Is Option "PageFlip" enabled? */
     Bool pageflip;
     Bool force_24_32;
@@ -182,6 +185,7 @@ typedef struct {
 
     drmmode_shadow_scanout_rec shadow_rotated;
     drmmode_shadow_scanout_ptr shadow_nonrotated;
+    drmmode_shadow_scanout_ptr shadow_nonrotated_back;
 
     PixmapPtr prime_pixmap;
     PixmapPtr prime_pixmap_back;
diff --git a/hw/xfree86/drivers/modesetting/modesetting.man b/hw/xfree86/drivers/modesetting/modesetting.man
index abf450c7c..67a5d8259 100644
--- a/hw/xfree86/drivers/modesetting/modesetting.man
+++ b/hw/xfree86/drivers/modesetting/modesetting.man
@@ -94,6 +94,14 @@ In an effort to reduce the performance cost, the latency may be negatively
 affected. The default is
 .B off.
 .TP
+.BI "Option \*qTearFree\*q \*q" boolean \*q
+Enable tearing prevention using the hardware page flipping mechanism. This
+option currently doesn't have any effect for rotated CRTCs. It requires
+allocating two separate scanout buffers for each non-rotated CRTC.
+.br
+This feature may reduce performance, increase latency and introduce stuttering. The default is
+.B off.
+.TP
 .SH "SEE ALSO"
 @xservername@(@appmansuffix@), @xconfigfile@(@filemansuffix@), Xserver(@appmansuffix@),
 X(@miscmansuffix@)
diff --git a/hw/xfree86/drivers/modesetting/pageflip.c b/hw/xfree86/drivers/modesetting/pageflip.c
index 027289f4e..407f524ec 100644
--- a/hw/xfree86/drivers/modesetting/pageflip.c
+++ b/hw/xfree86/drivers/modesetting/pageflip.c
@@ -356,4 +356,65 @@ error_out:
 #endif /* GLAMOR_HAS_GBM */
 }
 
+static void
+ms_scanout_flip_abort(void *data)
+{
+    drmmode_crtc_private_ptr drmmode_crtc = data;
+
+    if (drmmode_crtc->shadow_nonrotated_back)
+        drmmode_crtc->shadow_nonrotated_back->update_seq = 0;
+}
+
+static void
+ms_scanout_flip_handler(uint64_t frame, uint64_t usec, void *data)
+{
+    drmmode_crtc_private_ptr drmmode_crtc = data;
+    drmmode_shadow_scanout_ptr tmp;
+
+    if (!drmmode_crtc->shadow_nonrotated || !drmmode_crtc->shadow_nonrotated_back)
+        return;
+
+    /* the flip has happened, exchange the front and back buffers */
+    tmp = drmmode_crtc->shadow_nonrotated;
+    drmmode_crtc->shadow_nonrotated = drmmode_crtc->shadow_nonrotated_back;
+    drmmode_crtc->shadow_nonrotated_back = tmp;
+
+    /* allow the next update to queue a new flip */
+    ms_scanout_flip_abort(data);
+}
+
+void
+ms_do_tearfree_flip(xf86CrtcPtr crtc)
+{
+    drmmode_crtc_private_ptr drmmode_crtc = crtc->driver_private;
+    ScreenPtr screen = xf86ScrnToScreen(crtc->scrn);
+    uint32_t seq;
+
+    /* exit if the current crtc is off or not using tear-free */
+    if (!ms_crtc_on(crtc) || !drmmode_crtc->shadow_nonrotated_back)
+        return;
+
+    /* exit if an update has already been scheduled */
+    if (drmmode_crtc->shadow_nonrotated_back->update_seq)
+        return;
+
+    /* copy the screen pixmap to our current backbuffer */
+    drmmode_update_scanout_buffer(crtc,
+                                  drmmode_crtc->shadow_nonrotated_back);
+
+    seq = ms_drm_queue_alloc(crtc, drmmode_crtc, ms_scanout_flip_handler,
+                             ms_scanout_flip_abort);
+    if (!seq) {
+        xf86DrvMsg(crtc->scrn->scrnIndex, X_WARNING,
+                   "ms_drm_queue_alloc failed (%s). This may result in stale "
+                   "content displayed on the screen\n", strerror(errno));
+        return;
+    }
+
+    /* schedule a flip, and update the update_seq on success */
+    if (do_queue_flip_on_crtc(screen, crtc,
+                              drmmode_crtc->shadow_nonrotated_back->fb_id,
+                              DRM_MODE_PAGE_FLIP_EVENT, seq))
+        drmmode_crtc->shadow_nonrotated_back->update_seq = seq;
+}
 #endif
-- 
2.25.1

